core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -0,0 +1,183 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.util;
+
+import com.google.common.base.Supplier;
+
+import io.grpc.Attributes;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.ExperimentalApi;
+import io.grpc.LoadBalancer;
+import io.grpc.NameResolver;
+import io.grpc.ResolvedServerInfo;
+import io.grpc.Status;
+import io.grpc.TransportManager;
+import io.grpc.TransportManager.InterimTransport;
+import io.grpc.internal.RoundRobinServerList;
+
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.concurrent.GuardedBy;
+
+
+/**
+ * A {@link LoadBalancer} that provides round-robin load balancing mechanism over the
+ * addresses from the {@link NameResolver}.  The sub-lists received from the name resolver
+ * are considered to be an {@link EquivalentAddressGroup} and each of these sub-lists is
+ * what is then balanced across.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
+public final class RoundRobinLoadBalancerFactory extends LoadBalancer.Factory {
+
+  private static final RoundRobinLoadBalancerFactory instance = new RoundRobinLoadBalancerFactory();
+
+  private RoundRobinLoadBalancerFactory() {
+  }
+
+  public static RoundRobinLoadBalancerFactory getInstance() {
+    return instance;
+  }
+
+  @Override
+  public <T> LoadBalancer<T> newLoadBalancer(String serviceName, TransportManager<T> tm) {
+    return new RoundRobinLoadBalancer<T>(tm);
+  }
+
+  private static class RoundRobinLoadBalancer<T> extends LoadBalancer<T> {
+    private static final Status SHUTDOWN_STATUS =
+        Status.UNAVAILABLE.augmentDescription("RoundRobinLoadBalancer has shut down");
+
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private RoundRobinServerList<T> addresses;
+    @GuardedBy("lock")
+    private InterimTransport<T> interimTransport;
+    @GuardedBy("lock")
+    private Status nameResolutionError;
+    @GuardedBy("lock")
+    private boolean closed;
+
+    private final TransportManager<T> tm;
+
+    private RoundRobinLoadBalancer(TransportManager<T> tm) {
+      this.tm = tm;
+    }
+
+    @Override
+    public T pickTransport(Attributes affinity) {
+      synchronized (lock) {
+        if (closed) {
+          return tm.createFailingTransport(SHUTDOWN_STATUS);
+        }
+        if (addresses == null) {
+          if (nameResolutionError != null) {
+            return tm.createFailingTransport(nameResolutionError);
+          }
+          if (interimTransport == null) {
+            interimTransport = tm.createInterimTransport();
+          }
+          return interimTransport.transport();
+        }
+      }
+      return addresses.getTransportForNextServer();
```

Comment:
`addresses` should be protected by lock. You can assign `addresses` to a local variable inside `synchronized` and use the local variable here.